### PR TITLE
Datetime implementation for Mongo

### PR DESCRIPTION
**Pull request by timtuun**
You can query data based on datetime using mongo provider.
Supports most common iso8601 formats e.g.
datetime=2019-11-14T11:16
datetime=../2019
datetime=2020/..
...

Implementation is based on datetime range that is
parsed from url and passed to provider's query -method.

### CODE REVIEW

File: pygeoapi/util.py [hunk 1]
```diff
@@ -67,6 +68,28 @@ def dategetter(date_property, collection):
     return value.isoformat()
 
 
+def get_timezone_aware(d):
```
[the remark this answers is not in the document]
Comment (timtuun) [reply pygeoapi/util.py]: Fixed

File: tests/test_iso8601.py [hunk 2]
```diff
@@ -0,0 +1,128 @@
+# -*- coding: utf-8 -*-
```
[the remark this answers is not in the document]
Comment (timtuun) [reply tests/test_iso8601.py]: Fixed

File: pygeoapi/iso8601.py [hunk 3]
```diff
@@ -0,0 +1,192 @@
+
```
[the remark this answers is not in the document]
Comment (timtuun) [reply pygeoapi/iso8601.py]: Fixed

File: pygeoapi/iso8601.py [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+import re
```
[the remark this answers is not in the document]
Comment (timtuun) [reply pygeoapi/iso8601.py]: I renamed it to datetime_range.py to better descibe it's purpose.

File: pygeoapi/iso8601.py [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+        start = complete(start, end)
+        end = complete(end, start)
+
+    d_start = to_datetime(start) if start else None
+    d_end = to_datetime(end, ceil=True) if end else None
+
+    return DatetimeRange(d_start, d_end)
+
+
+__all__ = ['DatetimeRange', 'parse']
```
[the remark this answers is not in the document]
Comment (timtuun) [reply pygeoapi/iso8601.py]: Not really. Removed

File: pygeoapi/iso8601.py [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+
+
+# Internal parts
```
[the remark this answers is not in the document]
Comment (timtuun) [reply pygeoapi/iso8601.py]: Fixed

File: pygeoapi/util.py [hunk 7]
```diff
@@ -34,6 +34,7 @@
 import json
 import logging
 import os
+import pytz
```
[the remark this answers is not in the document]
Comment (timtuun) [reply pygeoapi/util.py]: Added

File: pygeoapi/iso8601.py [hunk 8]
```diff
@@ -0,0 +1,192 @@
+
+import re
+from datetime import datetime, timedelta, timezone
```
[the remark this answers is not in the document]
Comment (timtuun) [reply pygeoapi/iso8601.py]: I can add that, but I am not really sure what kind of logging would be feasible in this file.

[the remark this answers is not in the document]
Comment (timtuun) [reply pygeoapi/iso8601.py]: I added some simple logging. I hope that is sufficient

**Comment by timtuun** [comment]
Also, I found out that sortby -name was not in reserved_fieldnames -list and was not passed to providers, but caused and exception if added to url.

File: pygeoapi/datetime_range.py [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+
+def prefer(a, b):
+    return a if (a is not None) else b
+
+
+def intOrNone(a):
+    return int(a) if (a is not None) else None
+
+
+def match(regexp, s: str) -> tuple:
```
Comment (tomkralidis) [inline pygeoapi/datetime_range.py]: docstring

File: pygeoapi/datetime_range.py [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+    m = regexp.match(s)
+    if m is None:
+        raise ValueError('Invalid datetime string: {}'.format(s))
+    return m.groups()
+
+
+def tzinfo(s: str):
```
Comment (tomkralidis) [inline pygeoapi/datetime_range.py]: docstring

File: pygeoapi/datetime_range.py [hunk 11]
```diff
@@ -0,0 +1,219 @@
...
+        return None
+    elif s == 'Z':
+        return timezone.utc
+    else:
+        sign, hours, minutes = match(TIMEZONE_RE, s)
+        delta = timedelta(hours=int(hours), minutes=intOrNone(minutes) or 0)
+        return timezone(delta if (sign == '+') else -delta)
+
+
+def date_tuple(year, month, day) -> tuple:
```
Comment (tomkralidis) [inline pygeoapi/datetime_range.py]: docstring

File: pygeoapi/datetime_range.py [hunk 12]
```diff
@@ -0,0 +1,219 @@
...
+    return intOrNone(year), intOrNone(month), intOrNone(day)
+
+
+def time_tuple(hour, minute, second, fraction, tz) -> tuple:
```
Comment (tomkralidis) [inline pygeoapi/datetime_range.py]: docstring

File: pygeoapi/datetime_range.py [hunk 13]
```diff
@@ -0,0 +1,219 @@
...
+    digits = len(fraction[:6]) if fraction else None
+    microsecond = fraction[:6].ljust(6, '0') if fraction else None
+    return (intOrNone(hour), intOrNone(minute), intOrNone(second),
+            intOrNone(microsecond), digits, tzinfo(tz))
+
+
+def parse_datetime(s: str) -> Optional[tuple]:
```
Comment (tomkralidis) [inline pygeoapi/datetime_range.py]: docstring

File: pygeoapi/datetime_range.py [hunk 14]
```diff
@@ -0,0 +1,219 @@
...
+            time = time_tuple(*match(TIME_RE, s))
+            date = (None, None, None)
+        else:
+            date = date_tuple(*match(DATE_RE, s))
+            time = (None, None, None, None, None, None)
+
+    return date + time
+
+
+def complete(d0: tuple, d1: tuple) -> tuple:
```
Comment (tomkralidis) [inline pygeoapi/datetime_range.py]: update docstring with params and return

File: pygeoapi/datetime_range.py [hunk 15]
```diff
@@ -0,0 +1,219 @@
...
+    '''  Replace leading None values of d0 by corresponding value of d1'''
+    leading_nones = next(i for i, v in enumerate(d0) if v is not None)
+    return d1[:leading_nones] + d0[leading_nones:]
+
+
+def to_datetime(values: tuple, ceil=False) -> datetime:
```
Comment (tomkralidis) [inline pygeoapi/datetime_range.py]: update docstring with params and return

File: pygeoapi/provider/mongo.py [hunk 16]
```diff
@@ -63,21 +64,40 @@ def __init__(self, provider_def):
         dbclient = MongoClient(self.data)
         self.featuredb = dbclient.get_default_database()
         self.collection = provider_def['collection']
+        self.datetime_field = (provider_def['datetime_field']
+                               if 'datetime_field' in provider_def
+                               else 'datetime')
         self.featuredb[self.collection].create_index([("geometry", GEOSPHERE)])
+        self.fields = self.get_fields()
+
+    def _get_field_type(self, obj):
```
Comment (tomkralidis) [inline pygeoapi/provider/mongo.py]: Are you able to use https://github.com/geopython/pygeoapi/blob/master/pygeoapi/util.py#L70 in anyway instead?  No preference, just pointing out its existence which may or may be of use.

File: pygeoapi/util.py [hunk 17]
```diff
@@ -67,6 +68,36 @@ def dategetter(date_property, collection):
...
+    """
+    if isinstance(d, datetime):
+        return d
+    elif isinstance(d, date):
+        return datetime.combine(d, time.max if ceil else time.min)
+    else:
+        raise ValueError
+
+
 def get_typed_value(value):
```
Comment (tomkralidis) [inline pygeoapi/util.py]: or here? https://github.com/geopython/pygeoapi/blob/master/pygeoapi/util.py#L70

**Comment by tomkralidis** [comment]
@timtuun in addition, see https://github.com/geopython/pygeoapi/pull/337 to which you will need to rebase against, given updates in PyYAML to support (timely enough) timezone aware date parsing.

File: pygeoapi/api.py [hunk 18]
```diff
@@ -35,15 +35,15 @@
 import logging
 import urllib.parse
 
-from dateutil.parser import parse as dateparse
-
 from pygeoapi import __version__
+from pygeoapi.datetime_range import parse as datetimerange_parse
```
[the remark this answers is not in the document]
Comment (timtuun) [reply pygeoapi/api.py]: Fixed

File: pygeoapi/api.py [hunk 19]
```diff
@@ -460,7 +460,7 @@ def get_collection_items(self, headers, args, dataset, pathinfo=None):
 
         properties = []
         reserved_fieldnames = ['bbox', 'f', 'limit', 'startindex',
-                               'resulttype', 'datetime']
+                               'resulttype', 'datetime', 'sortby']
```
[the remark this answers is not in the document]
Comment (timtuun) [reply pygeoapi/api.py]: :)

File: pygeoapi/datetime_range.py [hunk 20]
```diff
@@ -0,0 +1,219 @@
...
+    '''
+    DatetimeRange contains date range with
+    start (inclusive)
+    end (exclusive)
+    datetimes.
+    E.g. 2007-03-01 will contain:
+    start = 2007-03-01
+    end = 2007-03-02
+    '''
+    def __init__(self, start: Optional[datetime], end: Optional[datetime]):
```
[the remark this answers is not in the document]
Comment (timtuun) [reply pygeoapi/datetime_range.py]: Added

File: pygeoapi/datetime_range.py [hunk 21]
```diff
@@ -0,0 +1,219 @@
...
+            return NotImplemented
+        return (self.start, self.end) == (other.start, other.end)
+
+    def __repr__(self):
+        start = self.start.isoformat() if (self.start is not None) else '..'
+        end = self.end.isoformat() if (self.end is not None) else '..'
+        return 'DatetimeRange({}/{})'.format(start, end)
+
+
+def parse(s: str) -> DatetimeRange:
```
[the remark this answers is not in the document]
Comment (timtuun) [reply pygeoapi/datetime_range.py]: Added

File: pygeoapi/datetime_range.py [hunk 22]
```diff
@@ -0,0 +1,219 @@
...
+    d_end = to_datetime(end, ceil=True) if end else None
+
+    return DatetimeRange(d_start, d_end)
+
+
+def prefer(a, b):
+    return a if (a is not None) else b
+
+
+def intOrNone(a):
```
[the remark this answers is not in the document]
Comment (timtuun) [reply pygeoapi/datetime_range.py]: Fixed

File: pygeoapi/datetime_range.py (hunk 13 above)
Comment (timtuun) [reply pygeoapi/datetime_range.py]: Added

File: pygeoapi/datetime_range.py (hunk 14 above)
Comment (timtuun) [reply pygeoapi/datetime_range.py]: Added

File: pygeoapi/datetime_range.py (hunk 15 above)
Comment (timtuun) [reply pygeoapi/datetime_range.py]: Added

File: pygeoapi/provider/mongo.py (hunk 16 above)
Comment (timtuun) [reply pygeoapi/provider/mongo.py]: I just tried with it, but tests crashed due to following error:
E           TypeError: argument of type 'int' is not iterable

pygeoapi/util.py:111: TypeError

...so I didn't change it (at least yet: )

File: pygeoapi/util.py (hunk 17 above)
Comment (timtuun) [reply pygeoapi/util.py]: See above

**Comment by timtuun** [comment]
> @timtuun in addition, see #337 to which you will need to rebase against, given updates in PyYAML to support (timely enough) timezone aware date parsing.

Yeah. I actually had already fixed it, with slightly different approach. :)

**Comment by tomkralidis** [comment]
@timtuun let us know when this PR is ready for review again.

**Comment by timtuun** [comment]
> 
> 
> @timtuun let us know when this PR is ready for review again.

@tomkralidis it is already ready for review. Sorry I wasn't clear about this in my previous messages. :)

**Comment by timtuun** [comment]
@tomkralidis did you notice, this pull request is ready for review?

**Comment by tomkralidis** [comment]
@timtuun re-testing Travis

**Comment by tomkralidis** [comment]
@timtuun thanks for this excellent implementation.  I've tried this locally and it turns out that even when dates are input via `datetime=`, the resulting objects are datetimes.  What if I have my data defined as dates and not datetimes?  What if my backend (of billions of rows) temporal data is stored as true date types?

I suppose one could test for incoming `datetime` values  `HH::MM:SS` being `00:00:00` and then `.date()` them.  I could do this in my provider backend, but would rather have this upstream so that all providers do not have to replicate this logic.

Should we instead have a `TemporalRange` object which supports any of `datetime`, `date`, `..` or `None` ?

**Comment by tomkralidis** [comment]
Or in `pygeoapi.api`, test for this and update the types accordingly before passing to a provider query.

P.S. feel free to drop into our [Gitter channel](https://gitter.im/geopython/pygeoapi) if you'd like

**Comment by timtuun** [comment]
> @timtuun thanks for this excellent implementation. I've tried this locally and it turns out that even when dates are input via `datetime=`, the resulting objects are datetimes. What if I have my data defined as dates and not datetimes? What if my backend (of billions of rows) temporal data is stored as true date types?
>
> I suppose one could test for incoming `datetime` values `HH::MM:SS` being `00:00:00` and then `.date()` them. I could do this in my provider backend, but would rather have this upstream so that all providers do not have to replicate this logic.
> 
> Should we instead have a `TemporalRange` object which supports any of `datetime`, `date`, `..` or `None` ?

If I understood correctly, I guess we have 2 options:
1) Improve the DatetimeRange.__contains__ method to accept date and None objects aswell and use that method. However, there are likely performance issues with that when you have large data.
2) Implement start_date() and end_date() -methods to DatetimeRange, so you can get date objects out from DatetimeRange. However, in this case we need to decide, what kind of date objects do we get from a query like datetime=2007-03-01T13:00:00/2008-05-11T15:30:00 ?

We could also implement both. :)

**Comment by tomkralidis** [comment]
cc @francbartoli 

Revisiting this PR. @timtuun what kind of performance issues would we find in option 1?  What if a user really wants to request a date (and not a datetime)?

**Comment by ancatfi** [comment]
> cc @francbartoli
> 
> Revisiting this PR. @timtuun what kind of performance issues would we find in option 1? What if a user really wants to request a date (and not a datetime)?

First I want to understand what a user really wants when he/she requests the date, e.g. 2020-04-27. For me it means any event in the range 2020-04-27T00:00:00/T23:59:59.999. So finally there's always some kind of datetime range with exact start and end points (unless the end point is left open).

About the performance: the db queries should be optimized so that most of the data is filtered as soon as possible by the db already, not by post filtering at python level. In case of datetime that means that we must know the exact range of the datetime when building the db query. Without this information we'll fetch lot of unnecessary data just to be ignored at python level, and that is really performance issues.

**Comment by francbartoli** [comment]
@timtuun @ancatfi I'm just wondering why should we need all that stuff for handling date/datetime when we could use robust packages like [pendulum](https://pendulum.eustace.io/) and [arrow](https://arrow.readthedocs.io/en/latest/). It looks like to DRY

**Pull request closed without merging**